### ISSUE TITLE
xDS resource patching

### ISSUE THREAD

**Issue by kyessenov** [envoyproxy/envoy#8400]
As a continuation of the incremental xDS theme, it would be good to provide a general ability for xDS servers to patch individual xDS resources. We have two use cases in mind:
- large endpoint lists in cluster load assignments;
- break-glass modifications applied to generic xDS responses;

The ability would require implementing an xDS-aware patch semantics in Envoy. This may need support from additional proto annotations to provide customizations to how lists and maps are combined (see https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/strategic-merge-patch.md, for example).

/cc @howardjohn @mandarjog

**Comment by htuch** [comment]
We have had similar discussions in the context of UDPA FWIW. The main objection is the client complexity; other clients such as gRPC, which might not share the use cases that Envoy has, should not have to take on the implementation complexity. If we can figure out a lightweight approach (lighterweight than the linked strategic merge patch, things could be more palatable.

**Comment by kyessenov** [comment]
Let me sketch out a minimalistic proposal here, and let me know if that looks like the right direction.

1. A patch is a tuple (extended field path, protobuf message, operation). Patches are applied in a sequence. Patches that don't match extended field paths are no-ops.

2. Extended field path is a sequence of one or more selectors starting from `resources` field in the response. The target of the field path is an embedded message field value (singular or one of repeated). The field path may not match, in which case the patch is a no-op. The field paths are resolved dynamically to handle the polymorphic `protobuf.Any`. Each selector is one of:
2a. `f` refers to a singular message field `f` in the parent message.
2b. `f[i]` refers to `i`-th value of a repeated message field `f` in the parent message. This supports negative indexes, where `-1` refers to the last element.
2c. `f[g = string_literal]` refers to a value in the repeated message field `f` for which field `g` has the specified literal value. As a special case, `f[string_literal]` refers to a map entry value for a map field `f` with the given key.

3. The protobuf message must be of the same message type as the value of the field path for a patch to apply.

4. There are four supported operations:
4a. `Merge`: Update the value of the field path with the given message by applying the standard protobuf merge operation: rewrite primitive values, recursively merge message values, and concatenate repeated values.
4b. `Delete`: Delete the referenced field value (clear it for an embedded message field, delete from repeated elements, or remove from a map). The provided message is ignored.
4c. `Prepend`: Only applicable if the field path value refers to a value in a repeated field. Insert the given value right before the referenced value.
4d. `Append`: Only applicable if the field path value refers to a value in a repeated field. Insert the given value right after the referenced value.

Examples:
1. LDS patch to update HTTP buffer filter config: `(resources[name=inbound].filter_chains[0].filters[name=HCM].typed_config.http_filters[name=envoy.filters.http.buffer], {max_request_bytes: 1024}, Merge)`
2. RDS patch to delete a wildcard virtual host:
`(resources[name=inbound].virtual_hosts[name=wildcard], {}, Delete)`
3. CDS patch to change http2 settings:
`(resources[name=xds], {http2_protocol_options: {}}, Merge)`
4. LDS patch to prepend a special HTTP filter in the filter chain:
`(resources[name=inbound].filter_chains[0].filters[name=HCM].typed_config.http_filters, {name="my_filter", typed_config = "..."}, Prepend)`

**Comment by mattklein123** [comment]
@kyessenov I would suggest that we have a short meeting to discuss this before we dive too far into the implementation details. As @htuch mentioned, we have actually already discussed the concept of merging quite a bit in the context of federation. I have gone back and forth several times on whether we should implement something like this or not (with my current thinking erring on the side of yes TBH), but I think we should discuss motivations first to make sure we are doing this for the right reasons.

**Comment by kyessenov** [comment]
SGTM, would be happy to chat more. CC @rshriram who worked on xDS patching as an Istio API.

**Comment by rshriram** [comment]
What is the rationale for pushing the patching mechanism to the data plane? There was a non trivial amount of struct/any business that we had to handle in the control plane in order to do the protobuf merging. This is still not complete but at the very least, we have more flexibility to do this in the control plane than in the data plane IMO. On top of this, we also have to handle validation issues (invalid listeners, etc.). 

Pushing the patching mechanism to the data plane is simply moving all these problems from Pilot to Envoy. In a scenario where we manage to cache these patched XDS resources, we have the benefit of doing the patching once per N proxies. We would lose that benefit as well if we do the patching in the data plane.

**Comment by kyessenov** [comment]
@rshriram Specifically for istio, we are trying to solve two problems:
- O(n^2) config size on the wire when changes are propagated linearly; for example, EDS payload being sent with 1,2,3, and up to n endpoints. This calls for an incremental EDS payload.
- Istio extensions applicable across control planes, e.g. I'd guess similar to federation. We would like to have a declarative description of where to inject Istio extensions that can be applied across multiple control planes (e.g. various versions of Istio Pilot). This is not a full replacement for EnvoyFilter since for patching, we don't assume any prior knowledge about the semantics of names or config layout.

**Comment by htuch** [comment]
Please include me in any meeting you folks setup, there are implications for UDPA that I'd like to get clear. I'd prefer that we make this technique as dead simple to implement as possible; if we have to write more than ~100 lines-of-code to do the merge, I worry that non-Envoy xDS/UDPA clients are going to have a tough time implementing the feature. CC @markdroth

**Comment by markdroth** [comment]
I'd like to be included in this discussion too.  I'm really concerned about the complexity and manageability of protobuf merging.

I do agree that we need a form of EDS that is incremental at the granularity of individual endpoints, and I've mentioned this to Harvey before.  But I think that use-case is something we can address more specifically via a separate first-class mechanism (e.g., maybe making each endpoint its own resource) rather than via a patching mechanism.

**Comment by howardjohn** [comment]
+1 for being able to send individual endpoints. Is this any different than delta xDS though?

I agree with @rshriram though for the most part, it seems this is better suited for the control plane to handle

**Comment by kyessenov** [comment]
I agree that simplicity is paramount. I think the real problem is that the config "modules" that Envoy provides (which is basically an individual xDS resource) does not line up well with the config intent. For example:
- an HTTP filter that needs metadata from virtual hosts and individual routes. Due to how config is spread out, that requires modifying LDS listener and RDS listener. RDS listener is a fairly structured patch.

It'd be great if we can change the "module" boundaries (e.g. subdivide endpoints object), but I'm concerned this is going to be an on-going problem for other use cases.

**Comment by lambdai** [comment]
@howardjohn  I think delta and incremental granular is per X in XDS where X has a name. So adding single cluster/listener/route is possible
For Eds the named resource is cluster, so we cannot yet add/delete single endpoint

**Comment by mattklein123** [comment]
cc @wgallagher @gkleiman

**Comment by htuch** [comment]
If we do want to specify overlay paths, one option we might want to think about is https://goessner.net/articles/JsonPath/. Basically, being able to describe a path from root to a subtree in a message. Proto3/JSON are canonically mapped, so it seems somewhat reusable.

**Comment by kyessenov** [comment]
JSON patch is a reasonable choice, but it has a set of problems:
1) ability to apply JSON patches on protobufs directly without JSON roudtrip (which is fairly expensive). There are subtle issues deleting field values, for example, that makes this fairly tricky (e.g. setting field to "null" in JSON does not translate well back to proto)
2) loss of a-priori knowledge about xDS structure (e.g. select by filter name, unpack Any)

**Comment by htuch** [comment]
In the context of https://github.com/envoyproxy/envoy/pull/9100#discussion_r349647950, the fact that there already exists a WKT for this, `FieldMask`, arose. See https://developers.google.com/protocol-buffers/docs/reference/csharp/class/google/protobuf/well-known-types/field-mask. 

It looks like we have reasonable support in the various protobuf languages for merging and masking with this, e.g. see https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.util.field_mask_util.

@markdroth would you be amenable towards merging if we stick with only the well-known types and standard library support provided in the protobuf libraries?

**Comment by kyessenov** [comment]
Field mask has an unfortunate limitation of not supporting repeated fields. E.g. we can only replace leaf lists AFAICT. What we would like is to de-couple filters from xDS resources, so that filters can be inserted separately from updating the entire monolithic listeners. This aligns better with the intent-driven objective, but it also requires an ability to insert into lists of filters. See https://raw.githubusercontent.com/istio/installer/master/istio-control/istio-discovery/templates/telemetryv2.yaml for an example of patch configuration.

**Comment by mattklein123** [comment]
I think we need to avoid merging lists IMO or this gets way too complicated. I think the filter case can/will be handled separately by allowing for a separate filter config delivery service. We already plan on implementing this.

**Comment by htuch** [comment]
Yep, agreed on the plan-of-record for filter split. For field masks, we could actually have repeated field masks possibly to workaround this limitation, but that loses some of the simplicity advantages.

**Comment by seflerZ** [comment]
I think the most urgent thing for xDS patch is EDS. Our organization suffered from large clusters a lot because we have over 10 thousand endpoints in some clusters. I would like to know that whether we've reached in common that this feature is desirable.

**Comment by howardjohn** [comment]
I would agree that EDS patches are the most important, and my understanding is is that this problem is not solved by incremental xds making it even more important

**Comment by mattklein123** [comment]
@seflerZ @howardjohn why doesn't incremental EDS not fix your case? Can't you just send incremental adds/removes?

**Comment by howardjohn** [comment]
Perhaps it does, I need to take a closer look at it. My understanding is that it does not support the level of granularity we want. If we can add/remove individual LbEndpoints https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/endpoint/endpoint.proto#envoy-api-msg-endpoint-lbendpoint then I think it would be perfect. My understanding was we still had to send a full ClusterLoadAssignment though

**Comment by mattklein123** [comment]
No, you should be able to add/remove individual LbEndpoints unless I am completely misunderstanding what we implemented which would be unfortunate. cc @htuch @lizan @fredlas @wgallagher

**Comment by htuch** [comment]
@mattklein123 the problem is folks want to add or remove individual endpoints within a cluster, rather than incrementally add/remove clusters. This is something we need to address in v4/UDPA, but I feel it might need a closer solution in v3.

IMHO at this point we have a bunch of solutions, many of which could work. We need to find someone to own the design (ensuring it aligns with UDPA) and implementation for v3. I can't commit to this as I'm working on v3 and then the UDPA story next year, so looking for a volunteer to work on this.

**Comment by mattklein123** [comment]
OK my mistake, bummer. Yeah we will need someone to drive this forward at a design level agreed.

**Comment by kyessenov** [comment]
@htuch This has significant impact on v3 general structure. With respect to EDS, we can approach it from two sides:
1. Allow generic EDS payload patch, also allows flipping flags in C/R/LDS
2. Create a new resource (endpoint slice?) that can be bound clusters many-to-one. This is kind of similar to filter config discovery service.
Depending which way we pick, it would be nice to align how we patch sub-messages here. E.g. it would be odd to allow patching EDS but then filter config needs a special discovery service.

**Comment by seflerZ** [comment]
@htuch I'm very interested in this and will to contribute (At lease on Pilot sides).
@kyessenov I prefer to pick the first approach because it generally solved the problem which means it can be further applied to other xDS resources. Adding a new resource(type-url) of course will be a easy approach but it is likely hard to be normalized.

**Comment by seflerZ** [comment]
@htuch I'm very interested in this and willing to contribute (At least on Pilot side).
@kyessenov I prefer to pick the second approach. The first sounds great but takes heavy design and implementation. Although it can be applied generally (to other xDS resources), I don't see any needs in LDS\CDS at present. The RDS introduced VHDS to incrementally add or remove route configurations so I think EDS can also follow this way.

**Comment by mattklein123** [comment]
I am in favor of developing a generic patching mechanism. This would also help with things like adjusting host weights. However, I feel pretty strongly that we should not allowing merging or altering lists. In previous discussions with @htuch the way we thought about handling this is to move to a model in which any patchable resource must be addressable through a map. I think we could make endpoints adhere to this model by moving to a list structure to a map structure where the key is an arbitrary host name.

With all of that said, I think it's also fine to pursue a different xDS API that would allow incremental endpoint delivery.

cc @wgallagher

**Comment by markdroth** [comment]
@htuch If we do have to support patching, then I do agree that using the existing protobuf mechanism for this makes sense.  But I'm still not enthusiastic about the complexity of putting this into the API, both in terms of (a) making the API easy for humans to understand and manage and (b) implementation complexity (for example, gRPC can't take a dependency on proto, so we'd probably wind up needing to reimplement this functionality anyway, even if we use the standard protobuf types to represent the mask).

**Comment by htuch** [comment]
I think the least controversial thing to do to unblock this, which also would align with UDPA, is to add in another xDS for individual endpoints (rather than cluster endpoints, which is what we have today). Call this Endpoint Group Discovery Service (EGDS). EDS now references resources in EGDS.

The trick to making this compatible with incremental xDS is to take the namespace concept from UDPA-TP and apply this to v3 xDS. Essentially, each locality/priority will point to a namespace in EGDS, and the individual resources will be individual endpoints.

**Comment by seflerZ** [comment]
Having EDS resources reference EDGS is great because many clusters share the same endpoints. For example, if we have a group of Pods providing two services, in traditional EDS way we would send twice the number of the size of the group. The redundant data grows larger when the number of services within the same Pod becomes larger. EDGS solved this.
 
So the number of endpoints in EGDS is fully configurable and it can be assigned to different EDS? To my understanding, to get fully incremental patch, we should set each EGDS to contain only 1 endpoint. If we have a cluster with 10000+ endpoints, we get 10000+ EGDS resources?

**Comment by hzxuzhonghu** [comment]
> we should set each EGDS to contain only 1 endpoint.

Nope, @seflerZ it is like endpointslice, it will contain a bunch of endpoints. Which can solve most use scalability case.

**Comment by seflerZ** [comment]
> > we should set each EGDS to contain only 1 endpoint.
> 
> Nope, @seflerZ it is like endpointslice, it will contain a bunch of endpoints. Which can solve most use scalability case.

Alright, so it is supposed to be a bunch. As for the size, no detail requirements. You can also configure it to be 1 if you like. Is my understanding correct?

**Comment by htuch** [comment]
EGDS could be a singleton or many or even empty, I don't see a need to restrict cardinality.

**Comment by howardjohn** [comment]
For what its worth, a similar API in k8s, https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/, allows the grouping to be any size and defaults to 100 endpoints per group. I think its pretty reasonable to allow it to be arbitrary sizes. If I were to implement this in a control plane that is backed by EndpointSlices, it would probably make most sense to just have a 1:1 mapping of EndpointSlices and EGDS resources to keep things simple.

**Comment by kyessenov** [comment]
To summarize, I think there's a desire to extend endpoints with endpoint groups:
```proto
// addition to EDS
message ClusterLoadAssignment {
  // references endpoint groups by name
  repeated string endpoint_groups = 6
}
// new xDS payload
message EndpointGroup {
  string name = 1;
  repeated endpoint.LocalityLbEndpoints endpoints = 2;
}
```

I would also like to propose to generalize the filter config discovery service to return a list of filters instead of a single one (and allow empty ones):

```proto
// additions to LDS
message Filter {
  // oneof with name/typed_config pair
  string filter_group = 5;
}
message HttpFilter {
  // oneof with name/typed_config pair
  string filter_group = 5;
}
// new xDS payload
message FilterConfig {
  // maps to a list of either Filter or HttpFilter
  repeated Filter filter_config = 1;
}
``` 

What do you think about the filter config xDS @htuch ? I can send a PR to sketch out API changes before locking down v3.

**Comment by markdroth** [comment]
@kyessenov Note that we should make sure to add `ConfigSource` messages for these cases, so that we have a way to tell the client which xds server to talk to for the new resource types.

**Comment by kyessenov** [comment]
@markdroth Ack, I'll apply the best practices with regards to the config source when I create a PR

**Comment by mattklein123** [comment]
@kyessenov +1 on the filter config stuff, but this issue is confusing enough as it is. Let's please split EGDS into a new issue, and cover FCDS in https://github.com/envoyproxy/envoy/issues/7867. Thanks.

**Comment by htuch** [comment]
@kyessenov I have feedback on these, but let's do this in the distinct issues. I definitely don't want to completely change how CLA works though, we should aim for a light tough here, the minimal change to existing structure to enable endpoint groups.

**Comment by lambdai** [comment]
I see no reason to adopt EndpointSlice or EndpointGroup concept in envoy. I view EndpointSlice a way of data normalization. It reduce the transport load if we always do transport at full.
However,  if our long term goal is to support incremental at endpoint level, the early normalization is bringing more pain.
We now have to resolve the complexity of migrating endpoint from group A to group B.

**Comment by seflerZ** [comment]
@mattklein123 I've create a issue [here](https://github.com/istio/istio/issues/19287) to discuss the EDGS issue.

@lambdai The problem is I don't see a wide need for context merging in other xDS (like CDS/LDS). If we construct a systematic and generic solution for merging, it should support patching and merging in all data types/paths as @kyessenov have described here:

> 1. A patch is a tuple (extended field path, protobuf message, operation). Patches are applied in a sequence. Patches that don't match extended field paths are no-ops.
> 2. Extended field path is a sequence of one or more selectors starting from `resources` field in the response. The target of the field path is an embedded message field value (singular or one of repeated). The field path may not match, in which case the patch is a no-op. The field paths are resolved dynamically to handle the polymorphic `protobuf.Any`. Each selector is one of:
>    2a. `f` refers to a singular message field `f` in the parent message.
>    2b. `f[i]` refers to `i`-th value of a repeated message field `f` in the parent message. This supports negative indexes, where `-1` refers to the last element.
>    2c. `f[g = string_literal]` refers to a value in the repeated message field `f` for which field `g` has the specified literal value. As a special case, `f[string_literal]` refers to a map entry value for a map field `f` with the given key.
> 3. The protobuf message must be of the same message type as the value of the field path for a patch to apply.
> 4. There are four supported operations:
>    4a. `Merge`: Update the value of the field path with the given message by applying the standard protobuf merge operation: rewrite primitive values, recursively merge message values, and concatenate repeated values.
>    4b. `Delete`: Delete the referenced field value (clear it for an embedded message field, delete from repeated elements, or remove from a map). The provided message is ignored.
>    4c. `Prepend`: Only applicable if the field path value refers to a value in a repeated field. Insert the given value right before the referenced value.
>    4d. `Append`: Only applicable if the field path value refers to a value in a repeated field. Insert the given value right after the referenced value.
> 
> Examples:
> 
> 1. LDS patch to update HTTP buffer filter config: `(resources[name=inbound].filter_chains[0].filters[name=HCM].typed_config.http_filters[name=envoy.filters.http.buffer], {max_request_bytes: 1024}, Merge)`
> 2. RDS patch to delete a wildcard virtual host:
>    `(resources[name=inbound].virtual_hosts[name=wildcard], {}, Delete)`
> 3. CDS patch to change http2 settings:
>    `(resources[name=xds], {http2_protocol_options: {}}, Merge)`
> 4. LDS patch to prepend a special HTTP filter in the filter chain:
>    `(resources[name=inbound].filter_chains[0].filters[name=HCM].typed_config.http_filters, {name="my_filter", typed_config = "..."}, Prepend)`

This will be a huge PR and will touch large number of files. I don't think it is worthwhile. 

If you talk about merging endpoints only, it touches EDS logic and lack of isolation, making xDS message even more complex while this complexity doesn't bring benefits whole xDS messages. So why should we dive into a stable process and make changes to it?

This feature is very urgent at present as I thought. We need to implement it quickly and dead simple. If we come into another better solution in the future, we can adopt to it quickly by deprecating the EGDS because it is a stand alone process.

**Comment by mattklein123** [comment]
For discussion, the one specific need that I know we must solve is to allow endpoint weights to be federated away from the endpoints themselves (so that we can build red-line testing and other systems orthogonal to the primary EDS discovery service). There are obviously a lot of different ways of doing this but I just want to make sure that need is known as we figure this whole issue out in Q1.

**Comment by kyessenov** [comment]
MessageDifferencer seems like a good library choice: it has an implementation in C++ https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.util.message_differencer#MessageDifferencer.SpecificField and golang/Java. It has capability to compute the exact operational diff, that we can use as input to inverse.

**Comment by htuch** [comment]
@kyessenov thanks, it's kind of nice having this capability across languages.

I'll reiterate @markdroth concern here though; if we can do this without needing to rely on protobuf library, we will make some client's lives easier (e.g. those working only with JSON, where protobuf isn't available). For the specific case of EDS, we should be able to make something like EGDS work I reckon.

**Comment by seflerZ** [comment]
@htuch @mattklein123 If we make EDS references multiple EGDS resources. Is that means we must enable delta xDS on Pilot side in order to update EGDS fragment incrementally?

**Comment by seflerZ** [comment]
@htuch @mattklein123 If we make EDS references multiple EGDS fragments. Is that means we must enable delta xDS on Pilot side in order to update EGDS resource itself incrementally?

**Comment by howardjohn** [comment]
I do not think we would. Say as have 10k endpoints and we make our groups 100 endpoints each. Anytime an endpoint changes in a group we just send those 100 endpoints in the group (cheap). anytime we add or remove a group we send the list of egds names (100 strings, which is cheap)

**Comment by seflerZ** [comment]
@howardjohn Thanks, I think so too.

**Comment by zyfjeff** [comment]
@howardjohn 
A single EGDS itself is fully updated, but delta updates need to be supported between multiple EGDS, such as adding an EG or deleting an EG.

**Comment by howardjohn** [comment]
@zyfjeff I don't see why a delta EGDS list is needed. The number of EGDS is going to be fairly small, and all you need to send is the name. Even if you had 100k endpoint groups (probably equal to 10 million endpoints, assuming 100 per group, which is an absurdly high amount), thats only sending 1mb of data or so. Its just sending a list of strings pointing to the endpoints, which is much cheaper than the full endpoint struct

**Comment by zyfjeff** [comment]
@howardjohn What if you add an EG? Or delete an EG? If you don’t do delta xDS, you need to push the full amount of EG information to Envoy, and then Envoy assembles these EGs. If you support delta xDS, you only need to push the EG to be added, and then Envoy increases EG is fine.

**Comment by hzxuzhonghu** [comment]
>Envoy does the calculation as full data, and the calculation overhead is still very large.

What kind of calc does envoy do? I understand it need to aggregate all the Endpoints, so in this case, it just needs to add endpoints to the endpoints pool by locality and sum up the lb weight.

**Comment by zyfjeff** [comment]
@hzxuzhonghu 
I'm not very clear about the details of the endpoint update,
But as far as I know, it is not only the weight calculation, but also an O (N) host comparison (similar to metadata, but also health check, etc.).

**Comment by seflerZ** [comment]
@zyfjeff The full EG push only would happen occasionally when cluster endpoints changed significantly. What @hzxuzhonghu means is that EGDS shards will only bring more calculation by "add endpoints to the endpoints pool by locality and sum up the lb weight", other operations you mentioned would also be done with or without EGDS.

**Comment by gengleilei** [comment]
@kyessenov Refer to your definition of EGDS, I have some different ideas, I hope to communicate with you, as follows:
> 
> ```protobuf
> // addition to EDS
> message ClusterLoadAssignment {
>   // references endpoint groups by name
>   EgdsConfig endpoint_groups = 6
> }
> message EgdsConfig {
>   api.v3alpha.core.ConfigSource config_source = 1;
>   repeated string endpoint_group_name = 2;
> }
> // new xDS payload
> message EndpointGroup {
>   string name = 1;
>   repeated endpoint.LocalityLbEndpoints endpoints = 2;
> }
> ```
I think there will only be one EGDS configuration in a cluster, and multiple EndpointGroup resources can be requested through one EGDS, so define endpoint_group_name as repeated, and remove the repeated definition of endpoint_groups.

**Comment by kyessenov** [comment]
@gengleilei I think it might be valuable to compose multiple EGDS config sources for a single cluster. I was thinking of some multi-cluster use-case with distinct control planes.

**Comment by gengleilei** [comment]
@kyessenov yes, It makes sense in some multi-cluster use-case with distinct control planes.
